Fix to keep stride in return_and_correct_aliasing()

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113280
* #113279
* #118405
* __->__ #117860

Fixes #117794

Fix tripped the assert here: https://github.com/pytorch/pytorch/blob/86dedebeafdd7b08d21432cebd7538437d3b7509/torch/utils/_python_dispatch.py#L216

From investigation: I found that functionalization of an in-place op (`mul_` in this test case) results in the strides of `TwoTensor`'s `a` / `b` components being mutated to be contiguous. This is not reflected in the outer tensor, causing the assert to be tripped.

After discussion with Brian, I address this in this PR by disallowing input mutations on non-contiguous tensor subclass inputs for now.

Differential Revision: [D54214617](https://our.internmc.facebook.com/intern/diff/D54214617)

jbschlosser committed:
Fix to keep stride in return_and_correct_aliasing()

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

TODO: investigate why the fix trips the assert mentioned in the issue

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/two_tensor.py:
```diff
@@ -15,16 +15,26 @@ def __new__(cls, a, b):
+
+        out = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls,
+            # Set clearly incorrect size / stride / storage offset here and set
```
we might want to leave the `TwoTensor` class alone so its implementation is "correct" (so we can use it for more generic subclass tests) - maybe add a new one-off subclass in that other test file?

jbschlosser replied:
I changed `TwoTensor` under the premise that its implementation is not correct, since metadata mutation of its components via functionalization is not reflected in the outer tensor.

That said, I see from the failing tests that `return_and_correct_aliasing()` does try to reflect metadata mutation from in-place views to the outer tensor, and the change I made to store 0 for the underlying size breaks this. So this isn't a complete fix atm. I'm unclear on whether the complete fix needs to touch `TwoTensor` specifically or not.

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

Fix tripped the assert here: https://github.com/pytorch/pytorch/blob/86dedebeafdd7b08d21432cebd7538437d3b7509/torch/utils/_python_dispatch.py#L216

From investigation: I found that functionalization of an in-place op (`mul_` in this test case) results in the strides of `TwoTensor`'s `a` / `b` components being mutated to be contiguous. This is not reflected in the outer tensor, causing the assert to be tripped.

To address this, I set the `dispatch_sizes_strides_policy` of `TwoTensor` so that sizes / strides / storage offset are always queried from the underlying components. Now the stride mutation of the inner tensors due to functionalization is properly reflected in the outer tensor.

This seems like a broader issue affecting subclasses whose sizes / strides are dependent on those of the inner tensors; for these to work properly with functionalization / PT2 as a whole, they probably want to follow this `dispatch_sizes_strides_policy="sizes"` pattern.

[ghstack-poisoned]

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

Fix tripped the assert here: https://github.com/pytorch/pytorch/blob/86dedebeafdd7b08d21432cebd7538437d3b7509/torch/utils/_python_dispatch.py#L216

From investigation: I found that functionalization of an in-place op (`mul_` in this test case) results in the strides of `TwoTensor`'s `a` / `b` components being mutated to be contiguous. This is not reflected in the outer tensor, causing the assert to be tripped.

After discussion with Brian, I address this by disallowing input mutations on non-contiguous tensor subclass inputs for now.

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/functional_utils.py:
```diff
@@ -134,7 +134,14 @@ def are_all_mutations_under_no_grad_or_inference_mode(t):
 # - check_only_storage_mutation=True: only return true if there was a storage mutation
 # - check_only_storage_mutation=Flse: return true if there was any metadata mutation (including a storage mutation)
-def has_metadata_mutation(f_arg, arg, *, check_only_storage_mutation: bool):
+def has_metadata_mutation(
```
Hmm - "was the tensor non-contiguous" is technically unrelated to whether or not there was a metadata mutation, so it's a bit confusing to put the logic here (in our case, we already ban metadata mutations on subclass inputs - we just want to additionally ban data-mutations on subclass inputs, when we detect that the subclass was (originally) non-contiguous.

bdhirsh replied:
Would a more targeted check like this work?
```
mutates_data = has_data_mutation(f_arg) # from existing code
if mutates_data and not arg.is_contiguous() and is_traceable_wrapper_subclass(arg):
    raise RuntimeError(...)
```

jbschlosser replied:
Yeah I definitely agree it's confusing. I was thinking we could count the stride mutation that results from functionalization as a "metadata mutation", but that's certainly an abuse of that terminology :p

> Would a more targeted check like this work?

It'll work in general, but I will likely have to revisit this for NJT since we do want to mutate "non-contiguous" inputs for KV-caching. I expect them to have contiguous strides but be non-contiguous in the "has holes" sense, so this check will be over-eager in failing those.

jbschlosser committed:
Update on "Fix to keep stride in return_and_correct_aliasing()"


Fixes #117794

Fix tripped the assert here: https://github.com/pytorch/pytorch/blob/86dedebeafdd7b08d21432cebd7538437d3b7509/torch/utils/_python_dispatch.py#L216

From investigation: I found that functionalization of an in-place op (`mul_` in this test case) results in the strides of `TwoTensor`'s `a` / `b` components being mutated to be contiguous. This is not reflected in the outer tensor, causing the assert to be tripped.

After discussion with Brian, I address this in this PR by disallowing input mutations on non-contiguous tensor subclass inputs for now.

[ghstack-poisoned]

bdhirsh reviewed (approved):
lgtm!

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -f "ignore pre-existing failure"

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54214617).
